### DISCUSSION TITLE
API call in Custom Sampler to suppress the activities controlled by custom feature toggle

### DISCUSSION THREAD

**Discussion by akorade** [open-telemetry/opentelemetry-dotnet#3044] [Q&A]
We have a requirement to have instrumentation controlled by a feature toggle. The plan is to use WCF, SQL auto instrumentation libraries and manual instrumentation as required but all controlled by a flag which resides in  the database. So we want to suppress the activities from being created and exported by OpenTelemetry collector. 
I was thinking to have that check inside the sampler rather that three different places (WCF Filter, Tracer API, SQL Filter which is not exposed yet). The API which retrieves this value calls database every 15 mins (caches and returns cached value). Any inputs on the approach?

**Comment by cijothomas** [comment]
Do you intent to drop telemetry in the sampler itself, based on whether its coming from WCF/SQL/Etc? 
The samplers only have access to the Span name, and initial attributes/tags, so need to 1st confirm if this information is sufficient to make that decision.

**Reply by akorade** [reply]
So the behavior I want here is that if the suppress flag in database is true I want to stop exporting and creating activities from all the registered sources (WCF\SQL\Manual etc.), So as such I Don't require any data from Sampler to make that decision. 

Below is the sampler example I was thinking of hooking up in Tracer Provider, thinking to provide a delegate which internally calls an API to check with Database 
``` csharp
 public class CustomSampler : Sampler
    {
        private readonly Func<bool> _suppressTelemetry;

        public CustomSampler(Func<bool> suppressTelemetry)
        {
            _suppressTelemetry = suppressTelemetry;
        }
        public override SamplingResult ShouldSample(in SamplingParameters samplingParameters)
        {
// Calls API which retrieves value from database.
            if (_suppressTelemetry())
            {
                return new SamplingResult(false);
            }
            return new SamplingResult(SamplingDecision.RecordAndSample);
        }
    }
```
Will this approach have any performance impact .
The only other option i have to 
a. utilize WCF filters for WCF traces
b. check before manual tracing in tracer API before creating any activity
c. expose filter from SQL auto instrumentation library just like one for wcf  and utilize it for check
